File: problems/postfixevaluation.py
class operation() :

	def __init__(self) : 
		self.array = []
	
	def push(self , value) :
		self.array.append(value)
	
	def pop(self) :

		if self.empty() :
			return
		self.array.pop()

	def peek(self) :
		return self.array[-1]
	
	def empty(self) :
		return len(self.array) == 0

stack = operation()

def priority(operator) :

	if operator == '+' or operator == '-' : return 1
	elif operator == '/' or operator == '*' : return 2
	elif operator == '^': return 3
	else : return 0

def digito4alpha(character) :

	if ((ord(character) >= 48 and ord(character) <= 57) or (ord(character) >= 65 and ord(character) <= 90) or (ord(character) >= 97 and ord(character) <= 122)) :
		return True
	
	else :
		return False

def math(firstoperand , secondoperand , symbol) :

	if symbol == '+' : return firstoperand + secondoperand
	elif symbol == '-' : return firstoperand - secondoperand
	elif symbol == '*' : return firstoperand * secondoperand
	elif symbol == '/' : return firstoperand / secondoperand
	else : return 0

def infix2postfix(expression) :

	output = ""

	for i in expression :

		if i == ' ' :
			continue

		elif i == '(' :
			stack.push(i)

		elif digito4alpha(i) :
			output += i

		elif i == ')' :

			while not stack.empty() and (stack.peek() != '(') :

				output += stack.peek()
				stack.pop()

			stack.pop()

		else :

			while not stack.empty() and priority(i) <= priority(stack.peek()) :

				output += stack.peek()
				stack.pop()

			stack.push(i)

	while not stack.empty() :

		output += stack.peek()
		stack.pop()

	return output


def evaluation(postfix) :

	for i in postfix :

		if i == ' ' :
			continue

		elif digito4alpha(i) :
			stack.push(int(i))

		else :

			secondoperand = stack.peek()
			stack.pop()
			firstoperand = stack.peek()
			stack.pop()
			result = math(firstoperand , secondoperand , i)
			stack.push(result)

	return stack.peek()

def main() :

	_expression = "(3+2)+7/2*((7+3)*2)"		# (3+2)+7/2*((7+3)*2)
	_postfix = infix2postfix(_expression)	# 32+72/73+2**+
	_value = evaluation(_postfix)			# 75.0 

	print("Infix   : " , _expression)
	print("Postfix : " , _postfix)
	print("Value   : " , _value)

if __name__ == '__main__':
	main()

File: problems/postfixevaluation/postfixevaluation.cpp

#include <iostream>

using namespace std ;

#define size 100

typedef struct operation

{
	float array[size] ;
	int top = -1 ;

	bool empty()

	{
		if (top == -1) return true ;
		else return false ;
	}

	void push(float value)

	{
		if (top == size -1) return ;
		top += 1 ;
		array[top] = value ;
	}

	void pop()

	{
		if (empty()) return ;
		top -= 1 ;
	}

	float peek()

	{
		return array[top] ;
	}

} stack ;


int digitalpha(int character)

{
	if ((character >= 48 && character <= 57) || (character >= 65 && character <= 90) || (character >= 97 && character <= 122)) return true ;
	else return false ;
}

int priority(char symbol)

{
	if (symbol == '+' || symbol == '-') return 1 ;
	else if (symbol == '/' || symbol == '*') return 2 ;
	else if (symbol == '^') return 3 ;
	else return 0 ;
}

stack s ;

string infixt2postfix(string infix)

{
	string output = "" ;

	for (int i = 0 ; i < infix.length() ; i++)

	{
		if (infix[i] == ' ') 
			continue ;

		else if (infix[i] == '(')
			
			s.push(infix[i]) ;

		else if(digitalpha(infix[i]) || digitalpha(infix[i]))
			
			output += infix[i] ;

		else if(infix[i] == ')')

		{
			while (s.peek() != '(')

			{
				output += s.peek() ;
				s.pop() ;
			}
			
			s.pop() ;

		}

		else

		{
			while(!s.empty() && priority(infix[i]) <= priority(s.peek()))

			{
				output += s.peek() ;
				s.pop() ;
			}

			s.push(infix[i]) ;
		}
	}

	while(!s.empty())

	{
		output += s.peek() ;
		s.pop() ;
	}

	return output ;
}

float math(float firstoperand , float secondoperand , char _oprator)

{
	if (_oprator == '+') return firstoperand + secondoperand ;
	else if (_oprator == '-') return firstoperand - secondoperand ;
	else if (_oprator == '*') return firstoperand * secondoperand ;
	else if (_oprator == '/') return firstoperand / secondoperand ;
	else return 0 ;
}

float evaluation(string _postfix)

{
	for (int i = 0 ; i < _postfix.length() ; i++)

	{
		if (_postfix[i] == ' ')
			continue ;

		if (digitalpha(_postfix[i]))
			s.push(_postfix[i] - '0') ;
		
		else

		{
			float secondoperand = s.peek() ;
			s.pop() ;
			float firstoperand = s.peek() ;
			s.pop() ;

			float result = math(firstoperand , secondoperand , _postfix[i]) ;

			s.push(result) ;
		}
	}

	return s.peek() ;
}


int main(void)

{	
		
	string infix = "(3+2)+7/2*((7+3)*2)" ;	
	string _postfix = infixt2postfix(infix) ;
	float postfixvalue = evaluation(_postfix) ;

	cout << "Infix   : " << infix << endl ;
	cout << "Postfix : " << _postfix << endl ;
	cout << "Value   : " << postfixvalue << endl ;
	return 0 ;

}
